packages/cipherstash-proxy/src/proxy/zerokms/zerokms.rs

```diff
@@ -0,0 +1,254 @@
+            .max_capacity((config.server.cipher_cache_size as u64) * SCOPED_CIPHER_SIZE as u64)
+            .time_to_live(Duration::from_secs(config.server.cipher_cache_ttl_seconds))
+            .build();
+
+        Ok(ZeroKms {
+            zerokms_client: Arc::new(zerokms_client),
+            cipher_cache,
+        })
+    }
+
+    /// Generate a cache key for the keyset identifier
+    fn cache_key_for_keyset(keyset_id: &Option<KeysetIdentifier>) -> String {
+        match keyset_id {
+            Some(id) => format!("{}", id.0),
+            None => "default".to_string(),
+        }
+    }
+
+    /// Initialize cipher using the stored zerokms_config, with async Moka caching and memory tracking
+    pub async fn init_cipher(
+        &self,
+        keyset_id: Option<KeysetIdentifier>,
+    ) -> Result<Arc<ScopedCipher>, Error> {
+        let cache_key = Self::cache_key_for_keyset(&keyset_id);
+
+        // Check cache first
+        if let Some(cached_cipher) = self.cipher_cache.get(&cache_key).await {
+            debug!(target: PROXY, msg = "Use cached ScopedCipher", ?keyset_id);
+            counter!(KEYSET_CIPHER_CACHE_HITS_TOTAL).increment(1);
+            return Ok(cached_cipher);
+        }
+
+        let zerokms_client = self.zerokms_client.clone();
+
+        debug!(target: PROXY, msg = "Initializing ZeroKMS ScopedCipher", ?keyset_id);
+
+        let identified_by = keyset_id.as_ref().map(|id| id.0.clone());
+
+        match ScopedCipher::init(zerokms_client, identified_by).await {
+            Ok(cipher) => {
+                let arc_cipher = Arc::new(cipher);
+
+                counter!(KEYSET_CIPHER_INIT_TOTAL).increment(1);
+
+                // Store in cache
+                self.cipher_cache
+                    .insert(cache_key, arc_cipher.clone())
+                    .await;
+
+                // Update pending tasks to get accurate cache statistics
+                self.cipher_cache.run_pending_tasks().await;
+
+                let entry_count = self.cipher_cache.entry_count();
+                let memory_usage_bytes = self.cipher_cache.weighted_size();
+
+                info!(msg = "Connected to ZeroKMS");
+                debug!(target: PROXY, msg = "ScopedCipher cached", ?keyset_id, entry_count, memory_usage_bytes);
+
+                Ok(arc_cipher)
+            }
+            Err(err) => {
+                debug!(target: PROXY, msg = "Error initializing ZeroKMS ScopedCipher", error = err.to_string());
+                warn!(msg = "Error initializing ZeroKMS", error = err.to_string());
+
+                match err {
+                    cipherstash_client::zerokms::Error::LoadKeyset(_) => {
+                        Err(EncryptError::UnknownKeysetIdentifier {
+                            keyset: keyset_id.map_or("default".to_string(), |id| id.to_string()),
+                        }
+                        .into())
+                    }
+                    cipherstash_client::zerokms::Error::Credentials(_) => {
```

Comment:
New error

Reply:
Note: 
An incorrect `client_id` causes a `LoadKeyset` error, not an authentication/credentials error.